File: manager/procd/pkg/process/process.go
package process

import (
	"fmt"
	"os"
	"sync"
	"syscall"
	"time"

	"github.com/creack/pty"
)

// ProcessType defines the type of process.
type ProcessType string

const (
	// ProcessTypeREPL represents a REPL process (Python, Node, Bash, Zsh, Ruby, Lua, PHP, R, Perl, etc.)
	ProcessTypeREPL ProcessType = "repl"
	// ProcessTypeCMD represents a one-time command execution (e.g., /bin/ls, /bin/cat)
	ProcessTypeCMD ProcessType = "cmd"
)

// OutputSource defines the source of process output.
type OutputSource string

const (
	OutputSourceStdout OutputSource = "stdout"
	OutputSourceStderr OutputSource = "stderr"
	OutputSourcePTY    OutputSource = "pty"
	OutputSourcePrompt OutputSource = "prompt"
)

// ProcessState represents the current state of a process.
type ProcessState string

const (
	ProcessStateCreated  ProcessState = "created"
	ProcessStateStarting ProcessState = "starting"
	ProcessStateRunning  ProcessState = "running"
	ProcessStatePaused   ProcessState = "paused"
	ProcessStateStopped  ProcessState = "stopped"
	ProcessStateKilled   ProcessState = "killed"
	ProcessStateCrashed  ProcessState = "crashed"
)

// PTYSize represents terminal dimensions.
type PTYSize struct {
	Rows uint16 `json:"rows"`
	Cols uint16 `json:"cols"`
}

// ProcessConfig holds configuration for creating a process.
type ProcessConfig struct {
	Type       ProcessType       `json:"type"`
	Language   string            `json:"language"` // For REPL: python, node, bash, zsh, ruby, lua, php, r, perl, etc.
	Code       string            `json:"code"`     // For REPL: code to execute
	Command    []string          `json:"command"`  // For CMD: command path and arguments, e.g., ["/bin/ls", "-la"]
	CWD        string            `json:"cwd"`
	EnvVars    map[string]string `json:"env_vars"`
	PTYSize    *PTYSize          `json:"pty_size"`
	Term       string            `json:"term"`
	BufferSize int               `json:"buffer_size"` // Number of messages to buffer for the output multiplexer, default is 64 if not set
}

// ProcessOutput represents output from a process.
type ProcessOutput struct {
	Source OutputSource `json:"source"`
	Data   []byte       `json:"data"`
}

// ExecutionResult represents the result of code/command execution.
type ExecutionResult struct {
	Output   []byte `json:"output"`
	Error    string `json:"error,omitempty"`
	ExitCode int    `json:"exit_code"`
}

// ResourceUsage represents resource consumption of a process.
// All fields are best-effort; unavailable stats will be zero or -1 for CPU percent.
type ResourceUsage struct {
	// Process-level stats (from /proc/[pid]/*)
	// CPUPercent is the CPU usage percentage since last sample.
	// Returns -1 on first call (no previous sample available).
	CPUPercent float64 `json:"cpu_percent"`
	// MemoryRSS is the Resident Set Size - actual physical memory used by process tree.
	MemoryRSS int64 `json:"memory_rss"`
	// MemoryVMS is the Virtual Memory Size - total virtual address space.
	MemoryVMS int64 `json:"memory_vms"`
	// OpenFiles is the number of open file descriptors.
	OpenFiles int `json:"open_files"`
	// ThreadCount is the number of threads in the process tree.
	ThreadCount int `json:"thread_count"`

	// Container-level stats (from cgroup, if available)
	// ContainerMemoryUsage is the total memory used by the container.
	ContainerMemoryUsage int64 `json:"container_memory_usage,omitempty"`
	// ContainerMemoryLimit is the memory limit for the container, 0 if unlimited.
	ContainerMemoryLimit int64 `json:"container_memory_limit,omitempty"`
	// ContainerMemoryWorkingSet is the non-reclaimable memory (used for OOM decisions).
	ContainerMemoryWorkingSet int64 `json:"container_memory_working_set,omitempty"`

	// I/O stats (from /proc/[pid]/io, may be unavailable)
	IOReadBytes  int64 `json:"io_read_bytes,omitempty"`
	IOWriteBytes int64 `json:"io_write_bytes,omitempty"`

	// Deprecated: Use MemoryRSS instead. Kept for backward compatibility.
	MemoryBytes int64 `json:"memory_bytes"`
}

// ExitEvent captures details about a process exit.
type ExitEvent struct {
	ProcessID     string
	ProcessType   ProcessType
	PID           int
	ExitCode      int
	Duration      time.Duration
	State         ProcessState
	StdoutPreview string
	StderrPreview string
	Config        ProcessConfig
}

// StartEvent captures details about a process start.
type StartEvent struct {
	ProcessID   string
	ProcessType ProcessType
	PID         int
	StartTime   time.Time
	State       ProcessState
	Config      ProcessConfig
}

// ExitHandler is a function that handles process exit events.
type ExitHandler func(ExitEvent)

// StartHandler is a function that handles process start events.
type StartHandler func(StartEvent)

// Process interface defines the contract for all process types.
type Process interface {
	// Identity
	ID() string
	Type() ProcessType
	PID() int

	// Lifecycle
	Start() error
	Stop() error
	Restart() error
	IsRunning() bool
	IsFinished() bool
	State() ProcessState
	AddStartHandler(StartHandler)
	AddExitHandler(ExitHandler)

	// Pause/Resume - sends SIGSTOP/SIGCONT to process group
	Pause() error
	Resume() error
	IsPaused() bool

	// I/O
	WriteInput(data []byte) error
	ReadOutput() <-chan ProcessOutput
	ResizePTY(size PTYSize) error
	SendSignal(sig syscall.Signal) error

	// Status
	ExitCode() (int, error)
	ResourceUsage() ResourceUsage
}

// BaseProcess provides common functionality for all process types.
type BaseProcess struct {
	id              string
	processType     ProcessType
	config          ProcessConfig
	state           ProcessState
	pid             int
	exitCode        int
	pty             *os.File
	outputMultiplex *MultiplexedChannel[ProcessOutput]
	startTime       time.Time

	// cpuTracker tracks CPU time between samples for percentage calculation
	cpuTracker *cpuTracker

	exitHandlers  []ExitHandler
	startHandlers []StartHandler

	mu sync.RWMutex

	inputQueue      chan []byte
	inputReady      chan struct{}
	inputStop       chan struct{}
	inputReadyOnce  sync.Once
	inputWriterOnce sync.Once
	inputStopOnce   sync.Once
}

// NewBaseProcess creates a new base process.
func NewBaseProcess(id string, processType ProcessType, config ProcessConfig) *BaseProcess {
	if config.BufferSize == 0 {
		config.BufferSize = 64
	}
	bp := &BaseProcess{
		id:              id,
		processType:     processType,
		config:          config,
		state:           ProcessStateCreated,
		outputMultiplex: NewMultiplexedChannel[ProcessOutput](config.BufferSize),
		cpuTracker:      newCPUTracker(),
		inputQueue:      make(chan []byte, config.BufferSize),
		inputReady:      make(chan struct{}),
		inputStop:       make(chan struct{}),
	}
	if processType != ProcessTypeREPL {
		close(bp.inputReady)
	}
	return bp
}

// AddExitHandler appends an exit handler to the handler chain.
// Handlers are executed in the order they were added.
func (bp *BaseProcess) AddExitHandler(handler ExitHandler) {
	bp.mu.Lock()
	defer bp.mu.Unlock()
	if handler != nil {
		bp.exitHandlers = append(bp.exitHandlers, handler)
	}
}

// AddStartHandler appends a start handler to the handler chain.
// Handlers are executed in the order they were added.
func (bp *BaseProcess) AddStartHandler(handler StartHandler) {
	bp.mu.Lock()
	defer bp.mu.Unlock()
	if handler != nil {
		bp.startHandlers = append(bp.startHandlers, handler)
	}
}

// ID returns the process ID.
func (bp *BaseProcess) ID() string {
	return bp.id
}

// Type returns the process type.
func (bp *BaseProcess) Type() ProcessType {
	return bp.processType
}

// PID returns the system process ID.
func (bp *BaseProcess) PID() int {
	bp.mu.RLock()
	defer bp.mu.RUnlock()
	return bp.pid
}

// State returns the current process state.
func (bp *BaseProcess) State() ProcessState {
	bp.mu.RLock()
	defer bp.mu.RUnlock()
	return bp.state
}

// IsRunning returns true if the process is running.
func (bp *BaseProcess) IsRunning() bool {
	bp.mu.RLock()
	defer bp.mu.RUnlock()
	return bp.state == ProcessStateRunning
}

// IsPaused returns true if the process is paused.
func (bp *BaseProcess) IsPaused() bool {
	bp.mu.RLock()
	defer bp.mu.RUnlock()
	return bp.state == ProcessStatePaused
}

func (bp *BaseProcess) IsFinished() bool {
	bp.mu.RLock()
	defer bp.mu.RUnlock()
	return bp.state == ProcessStateStopped || bp.state == ProcessStateKilled || bp.state == ProcessStateCrashed
}

// Pause sends SIGSTOP to the process group to pause the process and all its children.
func (bp *BaseProcess) Pause() error {
	bp.mu.Lock()
	defer bp.mu.Unlock()

	if bp.state != ProcessStateRunning {
		return ErrProcessNotRunning
	}

	if bp.pid <= 0 {
		return ErrProcessNotRunning
	}

	// Send SIGSTOP to the entire process group (negative PID)
	// This ensures all child processes are also paused
	if err := syscall.Kill(-bp.pid, syscall.SIGSTOP); err != nil {
		// If process group signal fails, try sending to the process directly
		if err := syscall.Kill(bp.pid, syscall.SIGSTOP); err != nil {
			return fmt.Errorf("%w: %v", ErrPauseFailed, err)
		}
	}

	bp.state = ProcessStatePaused
	return nil
}

// Resume sends SIGCONT to the process group to resume the process and all its children.
func (bp *BaseProcess) Resume() error {
	bp.mu.Lock()
	defer bp.mu.Unlock()

	if bp.state != ProcessStatePaused {
		return ErrProcessNotPaused
	}

	if bp.pid <= 0 {
		return ErrProcessNotRunning
	}

	// Send SIGCONT to the entire process group (negative PID)
	// This ensures all child processes are also resumed
	if err := syscall.Kill(-bp.pid, syscall.SIGCONT); err != nil {
		// If process group signal fails, try sending to the process directly
		if err := syscall.Kill(bp.pid, syscall.SIGCONT); err != nil {
			return fmt.Errorf("%w: %v", ErrResumeFailed, err)
		}
	}

	bp.state = ProcessStateRunning
	return nil
}

// ExitCode returns the process exit code.
func (bp *BaseProcess) ExitCode() (int, error) {
	bp.mu.RLock()
	defer bp.mu.RUnlock()
	return bp.exitCode, nil
}

// ResourceUsage returns resource usage statistics.
// This method reads from /proc and /sys/fs/cgroup pseudo-filesystems,
// which are available even in scratch containers as they are provided by the kernel.
// All stats are best-effort; failures are silently ignored and result in zero values.
func (bp *BaseProcess) ResourceUsage() ResourceUsage {
	bp.mu.RLock()
	pid := bp.pid
	bp.mu.RUnlock()

	usage := ResourceUsage{
		CPUPercent: -1, // Indicates no sample available
	}

	// If process is not running, return empty stats
	if pid <= 0 {
		return usage
	}

	// Aggregate stats for the process and all its children
	memStats, cpuStats, threads, openFiles, err := AggregateProcessTreeStats(pid)
	if err == nil {
		usage.MemoryRSS = memStats.RSS
		usage.MemoryVMS = memStats.VMS
		usage.MemoryBytes = memStats.RSS // Backward compatibility
		usage.ThreadCount = threads
		usage.OpenFiles = openFiles
	}

	// Calculate CPU percentage from delta since last sample
	if cpuStats != nil && bp.cpuTracker != nil {
		usage.CPUPercent = bp.cpuTracker.CalculateCPUPercent(cpuStats)
	}

	// Try to read I/O stats (may fail due to permissions)
	if ioStats, err := defaultProcReader.ReadIOStats(pid); err == nil {
		usage.IOReadBytes = ioStats.ReadBytes
		usage.IOWriteBytes = ioStats.WriteBytes
	}

	// Read container-level stats from cgroup (works in both runc and Kata)
	if containerMem, err := defaultCgroupReader.ReadContainerMemoryStats(); err == nil {
		usage.ContainerMemoryUsage = containerMem.Usage
		usage.ContainerMemoryLimit = containerMem.Limit
		usage.ContainerMemoryWorkingSet = containerMem.WorkingSet
	}

	return usage
}

// ReadOutput returns a channel for reading process output.
func (bp *BaseProcess) ReadOutput() <-chan ProcessOutput {
	ch, _ := bp.outputMultiplex.Fork()
	return ch
}

// WriteInput writes data to the process input.
func (bp *BaseProcess) WriteInput(data []byte) error {
	bp.mu.RLock()
	pty := bp.pty
	bp.mu.RUnlock()

	if len(data) == 0 {
		return nil
	}
	if pty == nil {
		return ErrProcessNotRunning
	}
	if bp.IsFinished() {
		return ErrProcessFinished
	}

	payload := make([]byte, len(data))
	copy(payload, data)

	select {
	case bp.inputQueue <- payload:
		return nil
	default:
		return ErrInputBufferFull
	}
}

// ResizePTY resizes the attached PTY, if present.
func (bp *BaseProcess) ResizePTY(size PTYSize) error {
	ptyFile := bp.GetPTY()
	if ptyFile == nil {
		return ErrPTYNotAvailable
	}

	if size.Rows == 0 || size.Cols == 0 {
		return fmt.Errorf("%w: rows and cols must be > 0", ErrInvalidPTYSize)
	}

	return pty.Setsize(ptyFile, &pty.Winsize{
		Rows: size.Rows,
		Cols: size.Cols,
	})
}

// SendSignal sends a signal to the process group, falling back to the PID.
func (bp *BaseProcess) SendSignal(sig syscall.Signal) error {
	bp.mu.RLock()
	pid := bp.pid
	state := bp.state
	bp.mu.RUnlock()

	if pid <= 0 || (state != ProcessStateRunning && state != ProcessStatePaused) {
		return ErrProcessNotRunning
	}

	if err := syscall.Kill(-pid, sig); err != nil {
		if err := syscall.Kill(pid, sig); err != nil {
			return fmt.Errorf("%w: %v", ErrSignalFailed, err)
		}
	}

	return nil
}

// SetState updates the process state.
func (bp *BaseProcess) SetState(state ProcessState) {
	bp.mu.Lock()
	defer bp.mu.Unlock()
	bp.state = state
}

// SetPTY sets the PTY file descriptor.
func (bp *BaseProcess) SetPTY(pty *os.File) {
	bp.mu.Lock()
	defer bp.mu.Unlock()
	bp.pty = pty
	bp.startInputWriter(pty)
}

func (bp *BaseProcess) signalInputReady() {
	bp.inputReadyOnce.Do(func() {
		close(bp.inputReady)
	})
}

// SignalInputReady marks the process as ready for input.
func (bp *BaseProcess) SignalInputReady() {
	bp.signalInputReady()
}

func (bp *BaseProcess) stopInputWriter() {
	bp.inputStopOnce.Do(func() {
		close(bp.inputStop)
	})
}

func (bp *BaseProcess) startInputWriter(pty *os.File) {
	bp.inputWriterOnce.Do(func() {
		go func() {
			select {
			case <-bp.inputReady:
			case <-bp.inputStop:
				return
			}
			for {
				select {
				case <-bp.inputStop:
					return
				case data := <-bp.inputQueue:
					if len(data) == 0 {
						continue
					}
					_, _ = pty.Write(data)
				}
			}
		}()
	})
}

// SetPID sets the system process ID.
func (bp *BaseProcess) SetPID(pid int) {
	bp.mu.Lock()
	defer bp.mu.Unlock()
	bp.pid = pid
}

// SetExitCode sets the exit code.
func (bp *BaseProcess) SetExitCode(code int) {
	bp.mu.Lock()
	defer bp.mu.Unlock()
	bp.exitCode = code
}

// SetStartTime records when the process starts.
func (bp *BaseProcess) SetStartTime(start time.Time) {
	bp.mu.Lock()
	defer bp.mu.Unlock()
	bp.startTime = start
}

// StartTime returns the recorded start time.
func (bp *BaseProcess) StartTime() time.Time {
	bp.mu.RLock()
	defer bp.mu.RUnlock()
	return bp.startTime
}

// NotifyExit invokes all exit handlers in the order they were added.
func (bp *BaseProcess) NotifyExit(event ExitEvent) {
	bp.mu.RLock()
	handlers := make([]ExitHandler, len(bp.exitHandlers))
	copy(handlers, bp.exitHandlers)
	bp.mu.RUnlock()

	for _, handler := range handlers {
		if handler != nil {
			handler(event)
		}
	}
}

// NotifyStart invokes all start handlers in the order they were added.
func (bp *BaseProcess) NotifyStart(event StartEvent) {
	bp.mu.RLock()
	handlers := make([]StartHandler, len(bp.startHandlers))
	copy(handlers, bp.startHandlers)
	bp.mu.RUnlock()

	for _, handler := range handlers {
		if handler != nil {
			handler(event)
		}
	}
}

// PublishOutput publishes output to all subscribers.
func (bp *BaseProcess) PublishOutput(output ProcessOutput) {
	bp.outputMultiplex.Publish(output)
}

// CloseOutput closes the output channel.
func (bp *BaseProcess) CloseOutput() {
	bp.outputMultiplex.Close()
}

// GetPTY returns the PTY file descriptor.
func (bp *BaseProcess) GetPTY() *os.File {
	bp.mu.RLock()
	defer bp.mu.RUnlock()
	return bp.pty
}

// GetConfig returns the process configuration.
func (bp *BaseProcess) GetConfig() ProcessConfig {
	return bp.config
}

File: manager/procd/src/process/process.js
import { spawn } from "node:child_process";
import crypto from "node:crypto";
import { EventEmitter } from "node:events";
import { accessSync, constants, existsSync } from "node:fs";
import path from "node:path";
import readline from "node:readline";
import { mergeEnvironment } from "../util/env.js";
import { errors } from "./errors.js";
import { normalizeReplConfig, defaultReadyToken } from "./repl-config.js";
import { processResourceUsage } from "./resource.js";

const maxBufferedOutput = 1 << 20;
const defaultPath = "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin";

function contextID() {
  return "ctx-" + cryptoRandom(8);
}

function cryptoRandom(n) {
  return crypto.randomBytes(Math.ceil(n / 2)).toString("hex").slice(0, n);
}

function appendLimited(current, data) {
  const next = current + data;
  if (Buffer.byteLength(next) <= maxBufferedOutput) return next;
  return next.slice(next.length - maxBufferedOutput);
}

function signalProcessGroup(pid, signal) {
  try {
    process.kill(-pid, signal);
  } catch (err) {
    if (err.code === "ESRCH") return;
    throw err;
  }
}

function executableExists(command, env) {
  if (!command) return false;
  const candidates = command.includes("/")
    ? [command]
    : String(env.PATH || defaultPath).split(":").filter(Boolean).map((dir) => path.join(dir, command));
  for (const candidate of candidates) {
    try {
      accessSync(candidate, constants.X_OK);
      return true;
    } catch {
      // Try the next PATH entry.
    }
  }
  return false;
}

function selectCandidate(candidates, env) {
  for (const candidate of candidates || []) {
    if (executableExists(candidate.name, env)) return candidate;
  }
  return candidates?.[0] || null;
}

function ptyHelperCommand() {
  const helperPath = process.env.SANDBOX0_PTY_HELPER_PATH || "/procd/bin/pty-helper";
  if (process.env.SANDBOX0_PTY_HELPER_USE_LOADER !== "true") return { command: helperPath, args: [] };
  const loaderPath = process.env.SANDBOX0_PROCD_LOADER_PATH || "/procd/runtime/ld-linux";
  const libraryPath = process.env.SANDBOX0_PROCD_LIBRARY_PATH || "/procd/runtime/lib";
  if (existsSync(loaderPath)) {
    return { command: loaderPath, args: ["--library-path", libraryPath, helperPath] };
  }
  return { command: helperPath, args: [] };
}

function encodePTYConfig(config) {
  return Buffer.from(JSON.stringify(config)).toString("base64");
}

export class ManagedProcess extends EventEmitter {
  constructor(config, sandboxEnv = {}) {
    super();
    this.id = config.id || contextID();
    this.type = config.type;
    this.alias = config.alias || "";
    this.command = [...(config.command || [])];
    this.cwd = config.cwd || "";
    this.envVars = { ...(config.envVars || {}) };
    this.ptySize = config.ptySize || null;
    this.term = config.term || "";
    this.replConfig = config.replConfig || null;
    this.sandboxEnv = sandboxEnv || {};
    this.state = "created";
    this.pid = 0;
    this.helperPID = 0;
    this.exitCode = null;
    this.createdAt = new Date();
    this.startedAt = null;
    this.finishedAt = null;
    this.paused = false;
    this.child = null;
    this.usesPTY = false;
    this.readyToken = defaultReadyToken;
    this.readyMode = "";
    this.readyBuffer = "";
    this.finished = false;
    this.stdout = "";
    this.stderr = "";
  }

  start() {
    if (this.isRunning()) throw errors.processAlreadyRunning();
    const spawnConfig = this.spawnConfig();
    this.state = "starting";
    this.finished = false;
    if (spawnConfig.pty) {
      this.startPTY(spawnConfig);
      return;
    }
    this.startDirect(spawnConfig);
  }

  startDirect(spawnConfig) {
    const child = spawn(spawnConfig.command[0], spawnConfig.command.slice(1), {
      cwd: spawnConfig.cwd || undefined,
      env: spawnConfig.env,
      detached: true,
      stdio: ["pipe", "pipe", "pipe"]
    });
    this.child = child;
    this.pid = child.pid || 0;
    this.startedAt = new Date();

    child.stdout.on("data", (chunk) => this.handleOutput("stdout", chunk));
    child.stderr.on("data", (chunk) => this.handleOutput("stderr", chunk));
    child.on("spawn", () => {
      this.state = "running";
      this.emit("start", this.startEvent());
      if (this.type === "repl") this.scheduleReplReady(spawnConfig.replConfig);
    });
    child.on("error", (err) => {
      this.state = "crashed";
      this.emit("error", err);
      this.finishProcess(1, "");
    });
    child.on("exit", (code, signal) => {
      this.finishProcess(code === null ? 137 : code, signal || "");
    });
  }

  startPTY(spawnConfig) {
    const helper = ptyHelperCommand();
    const helperEnv = mergeEnvironment({
      SANDBOX0_PTY_CONFIG: encodePTYConfig({
        command: spawnConfig.command,
        cwd: spawnConfig.cwd,
        env: spawnConfig.env,
        rows: spawnConfig.ptySize?.rows || 100,
        cols: spawnConfig.ptySize?.cols || 500
      })
    });
    const child = spawn(helper.command, helper.args, {
      env: helperEnv,
      stdio: ["pipe", "pipe", "pipe"]
    });
    this.child = child;
    this.usesPTY = true;
    this.helperPID = child.pid || 0;
    this.startedAt = new Date();

    const lines = readline.createInterface({ input: child.stdout });
    lines.on("line", (line) => this.handlePTYFrame(line, spawnConfig));
    child.stderr.on("data", (chunk) => this.handleOutput("stderr", chunk));
    child.on("error", (err) => {
      this.state = "crashed";
      this.emit("error", err);
      this.finishProcess(1, "");
    });
    child.on("exit", (code, signal) => {
      if (!this.finished) this.finishProcess(code === null ? 137 : code, signal || "");
    });
  }

  spawnConfig() {
    if (this.type === "cmd") {
      if (this.command.length === 0) throw errors.invalidCommand("invalid command: command is required");
      return {
        command: this.command,
        cwd: this.cwd,
        env: this.cmdEnvironment(),
        pty: Boolean(this.ptySize),
        ptySize: this.ptySize
      };
    }
    if (this.type === "repl") {
      const replConfig = normalizeReplConfig(this.alias || "python", this.replConfig);
      if (!replConfig) throw errors.unsupportedLanguage();
      const envLayer = this.replEnvironment(replConfig);
      const env = mergeEnvironment(this.sandboxEnv, this.envVars, envLayer);
      const candidate = selectCandidate(replConfig.candidates, env);
      if (!candidate) throw errors.unsupportedLanguage();
      this.alias = replConfig.name;
      this.readyMode = replConfig.ready?.mode || "";
      this.readyToken = replConfig.ready?.token || defaultReadyToken;
      this.readyBuffer = "";
      const ptySize = this.ptySize || { rows: 100, cols: 500 };
      return {
        command: [candidate.name, ...(candidate.args || [])],
        cwd: this.cwd,
        env,
        replConfig,
        pty: true,
        ptySize
      };
    }
    throw errors.unsupportedProcessType();
  }

  cmdEnvironment() {
    const env = mergeEnvironment(this.sandboxEnv, this.envVars);
    if (this.ptySize && !env.TERM) env.TERM = this.term || "xterm-256color";
    return env;
  }

  replEnvironment(replConfig) {
    const envLayer = {};
    for (const item of replConfig.env || []) {
      if (item.valueFrom === "term") envLayer[item.name] = this.term || replConfig.defaultTerm || "xterm-256color";
      else if (item.valueFrom === "prompt") envLayer[item.name] = replConfig.prompt?.customPrompt || defaultReadyToken;
      else envLayer[item.name] = item.value || "";
    }
    return envLayer;
  }

  scheduleReplReady(replConfig) {
    if (!replConfig) return;
    if (replConfig.ready?.mode === "startup_delay") {
      setTimeout(() => this.emitOutput("prompt", replConfig.ready?.token || defaultReadyToken), replConfig.ready?.startupDelayMs || 200);
    }
  }

  handleOutput(source, chunk) {
    const data = chunk.toString();
    if (source === "stdout") this.stdout = appendLimited(this.stdout, data);
    if (source === "stderr") this.stderr = appendLimited(this.stderr, data);
    this.emitOutput(source, data);
    if (this.type === "repl" && this.readyMode === "prompt_token") {
      if (this.detectReadyToken(data)) this.emitOutput("prompt", this.readyToken);
    }
  }

  handlePTYFrame(line, spawnConfig) {
    let frame;
    try {
      frame = JSON.parse(line);
    } catch {
      this.handleOutput("stderr", Buffer.from(line + "\n"));
      return;
    }
    if (frame.type === "start") {
      this.pid = frame.pid || 0;
      this.state = "running";
      this.emit("start", this.startEvent());
      if (this.type === "repl") this.scheduleReplReady(spawnConfig.replConfig);
      return;
    }
    if (frame.type === "output") {
      this.handleOutput("pty", Buffer.from(frame.data || "", "base64"));
      return;
    }
    if (frame.type === "error") {
      const message = frame.message || "pty helper failed";
      this.handleOutput("stderr", Buffer.from(message + "\n"));
      this.emit("error", new Error(message));
      return;
    }
    if (frame.type === "exit") {
      this.finishProcess(frame.exit_code ?? 0, frame.signal || "");
    }
  }

  detectReadyToken(data) {
    const token = this.readyToken || defaultReadyToken;
    if (!token) return false;
    const combined = this.readyBuffer + data;
    if (combined.includes(token)) {
      this.readyBuffer = "";
      return true;
    }
    this.readyBuffer = token.length > 1 ? combined.slice(-(token.length - 1)) : "";
    return false;
  }

  finishProcess(code, signal) {
    if (this.finished) return;
    this.finished = true;
    this.exitCode = code === null || code === undefined ? 137 : code;
    this.finishedAt = new Date();
    if (signal === "SIGKILL" || signal === "SIGTERM" || this.exitCode === 137 || this.exitCode === 143) this.state = "killed";
    else if (this.exitCode === 0) this.state = "stopped";
    else this.state = "crashed";
    this.emit("exit", this.exitEvent());
    this.emit("outputClose");
  }

  emitOutput(source, data) {
    const payload = { source, data };
    this.emit("output", payload);
    if (process.env.SANDBOX0_PROCESS_LOGS !== "false") {
      process.stdout.write(JSON.stringify({
        message: "sandbox process output",
        process_id: this.id,
        process_type: this.type,
        source,
        data
      }) + "\n");
    }
  }

  writeInput(data) {
    if (!this.child || !this.child.stdin || this.isFinished()) throw errors.processFinished();
    if (!this.isRunning()) throw errors.processNotRunning();
    if (this.usesPTY) {
      this.writePTYControl({ type: "input", data: Buffer.from(data).toString("base64") });
      return;
    }
    if (!this.child.stdin.write(data)) {
      // Backpressure is handled by Node; the API remains accepted.
    }
  }

  stop() {
    if (!this.isRunning()) return;
    if (this.usesPTY) this.writePTYControl({ type: "signal", signal: "SIGTERM" });
    else signalProcessGroup(this.pid, "SIGTERM");
    setTimeout(() => {
      if (!this.isRunning()) return;
      if (this.usesPTY) {
        this.writePTYControl({ type: "signal", signal: "SIGKILL" });
        if (this.child && !this.child.killed) this.child.kill("SIGKILL");
      } else {
        signalProcessGroup(this.pid, "SIGKILL");
      }
    }, 5000).unref();
  }

  restart() {
    this.stop();
    this.state = "created";
    this.paused = false;
    this.exitCode = null;
    this.child = null;
    this.start();
  }

  pause() {
    if (!this.isRunning()) throw errors.processNotRunning();
    if (this.usesPTY) this.writePTYControl({ type: "signal", signal: "SIGSTOP" });
    else signalProcessGroup(this.pid, "SIGSTOP");
    this.paused = true;
    this.state = "paused";
  }

  resume() {
    if (!this.child || this.isFinished()) throw errors.processNotRunning();
    if (this.usesPTY) this.writePTYControl({ type: "signal", signal: "SIGCONT" });
    else signalProcessGroup(this.pid, "SIGCONT");
    this.paused = false;
    this.state = "running";
  }

  sendSignal(signal) {
    if (!this.child || this.isFinished()) throw errors.processNotRunning();
    try {
      if (this.usesPTY) this.writePTYControl({ type: "signal", signal });
      else signalProcessGroup(this.pid, signal);
    } catch (err) {
      throw errors.signalFailed(err.message);
    }
  }

  resizePTY(size) {
    if (!this.usesPTY) throw errors.ptyUnavailable();
    this.writePTYControl({ type: "resize", rows: size.rows, cols: size.cols });
  }

  writePTYControl(frame) {
    if (!this.child || !this.child.stdin || this.isFinished()) throw errors.processFinished();
    if (!this.child.stdin.write(JSON.stringify(frame) + "\n")) {
      // The helper stdin pipe handles backpressure asynchronously.
    }
  }

  isRunning() {
    return this.state === "running" || this.state === "paused";
  }

  isPaused() {
    return this.paused || this.state === "paused";
  }

  isFinished() {
    return ["stopped", "killed", "crashed"].includes(this.state);
  }

  resourceUsage() {
    return processResourceUsage(this.pid);
  }

  startEvent() {
    return {
      process_id: this.id,
      process_type: this.type,
      pid: this.pid,
      command: this.command,
      env_vars: this.envVars,
      cwd: this.cwd,
      alias: this.alias
    };
  }

  exitEvent() {
    const duration = this.startedAt ? Date.now() - this.startedAt.getTime() : 0;
    return {
      process_id: this.id,
      process_type: this.type,
      pid: this.pid,
      exit_code: this.exitCode ?? 0,
      duration_ms: duration,
      stdout_preview: this.stdout.slice(0, 2048),
      stderr_preview: this.stderr.slice(0, 2048),
      state: this.state
    };
  }
}
